src/core/Akka.Persistence/Serialization/PersistenceMessageSerializer.cs

```diff
@@ -9,30 +9,32 @@
 using System.Collections.Generic;
 using System.Collections.Immutable;
 using System.Reflection;
+using System.Runtime.Serialization;
 using Akka.Actor;
 using Akka.Persistence.Fsm;
 using Akka.Persistence.Serialization.Proto.Msg;
 using Akka.Serialization;
+using Akka.Util;
 using Google.Protobuf;
 
 namespace Akka.Persistence.Serialization
 {
-    public class PersistenceMessageSerializer : Serializer
+    public sealed class PersistenceMessageSerializer : Serializer
     {
         public PersistenceMessageSerializer(ExtendedActorSystem system) : base(system)
         {
-            IncludeManifest = true;
         }
 
-        public override bool IncludeManifest { get; }
+        public override bool IncludeManifest { get; } = true;
 
         public override byte[] ToBinary(object obj)
         {
-            if (obj is IPersistentRepresentation) return GetPersistentMessage((IPersistentRepresentation)obj).ToByteArray();
-            if (obj is AtomicWrite) return GetAtomicWrite((AtomicWrite)obj).ToByteArray();
-            if (obj is AtLeastOnceDeliverySnapshot) return GetAtLeastOnceDeliverySnapshot((AtLeastOnceDeliverySnapshot)obj).ToByteArray();
-            if (obj is PersistentFSM.StateChangeEvent) return GetStateChangeEvent((PersistentFSM.StateChangeEvent)obj).ToByteArray();
-            if (obj.GetType().GetGenericTypeDefinition() == typeof(PersistentFSM.PersistentFSMSnapshot<>)) return GetPersistentFSMSnapshot(obj).ToByteArray();
+            if (obj is IPersistentRepresentation repr) return GetPersistentMessage(repr).ToByteArray();
+            if (obj is AtomicWrite aw) return GetAtomicWrite(aw).ToByteArray();
+            if (obj is AtLeastOnceDeliverySnapshot snap) return GetAtLeastOnceDeliverySnapshot(snap).ToByteArray();
+            if (obj is PersistentFSM.StateChangeEvent stateEvent) return GetStateChangeEvent(stateEvent).ToByteArray();
+            if (obj.GetType().GetTypeInfo().IsGenericType 
+                && obj.GetType().GetGenericTypeDefinition() == typeof(PersistentFSM.PersistentFSMSnapshot<>)) return GetPersistentFSMSnapshot(obj).ToByteArray();
 
             throw new ArgumentException($"Can't serialize object of type [{obj.GetType()}] in [{GetType()}]");
```

Comment:
Should this also be updated to SerializationException?

Reply:
good point

Reply:
I sync it with JVM version. They have `NonSerializableException` in `FromBinary` and they don't treat this exception as a fatal